src/transformers/models/llama/modeling_llama.py

```diff
@@ -92,61 +93,90 @@ def forward(self, hidden_states):
+        self.rope_config["rope_type"] = self.rope_config.get("rope_type", "default")
+
+        inv_freq = compute_frequencies(self.rope_config, device)
         self.register_buffer("inv_freq", inv_freq, persistent=False)
-        # For BC we register cos and sin cached
         self.max_seq_len_cached = max_position_embeddings
 
+        # BC: dynamic NTK and yarn used `scaling_factor` as a frequency parameter, not for position_ids scaling
+        if "dynamic" in self.rope_config["rope_type"] or "yarn" in self.rope_config["rope_type"]:
+            self.rope_config["scaling_factor"] = 1.0
+        # Special case: on yarn, `attention_factor` has a default suggested by the paper
+        if "yarn" in self.rope_config["rope_type"]:
+            self.rope_config["attention_factor"] = self.rope_config.get(
+                "attention_factor", 0.1 * math.log(scaling_factor) + 1.0
+            )
+
+    def dynamic_frequency_update(self, position_ids, device):
+        """dynamic RoPE layers need to recompute `inv_freq` when going beyond the original maximum sequence length"""
+        seq_len = torch.max(position_ids) + 1
+        if seq_len > self.max_seq_len_cached:
+            inv_freq = compute_frequencies(self.rope_config | {"seq_len": seq_len}, device)
+            self.register_buffer("inv_freq", inv_freq, persistent=False)  # TODO joao: may break with compilation
+            self.max_seq_len_cached = seq_len
+
     @torch.no_grad()
     def forward(self, x, position_ids):
-        # x: [bs, num_attention_heads, seq_len, head_size]
+        if "dynamic" in self.rope_config["rope_type"]:
+            self.dynamic_frequency_update(position_ids, device=x.device)
+
+        # Core RoPE block
+        if self.rope_config["scaling_factor"] != 1.0:
+            position_ids = position_ids.float() / self.rope_config["scaling_factor"]
         inv_freq_expanded = self.inv_freq[None, :, None].float().expand(position_ids.shape[0], -1, 1)
         position_ids_expanded = position_ids[:, None, :].float()
-        # Force float32 since bfloat16 loses precision on long contexts
-        # See https://github.com/huggingface/transformers/pull/29285
+        # Force float32 (see https://github.com/huggingface/transformers/pull/29285)
         device_type = x.device.type
         device_type = device_type if isinstance(device_type, str) and device_type != "mps" else "cpu"
         with torch.autocast(device_type=device_type, enabled=False):
             freqs = (inv_freq_expanded.float() @ position_ids_expanded.float()).transpose(1, 2)
             emb = torch.cat((freqs, freqs), dim=-1)
             cos = emb.cos()
             sin = emb.sin()
+
+        # Advanced RoPE types (e.g. yarn) apply a post-processing scaling factor, equivalent to scaling attention
+        if self.rope_config["attention_factor"] is not None:
+            cos = cos * self.rope_config["attention_factor"]
+            sin = sin * self.rope_config["attention_factor"]
```

Comment:
if this lives in a config vs in a tensor or buffer we will have device issue + we have less freedom IMO and no idea about the dtype